### PR TITLE
Adds support for Disambiguation using on context similarities based on vector models (now with extra rebase!)

### PR DESCRIPTION
**Pull request by phdowling**
--- Fixed version of #368, based of #369, removing duplicate commits and such ---
This is the PR for the first part of my GSoC project. This branch will only add the basic support for vector based contextsimilarities, the PR adding a more sophisticated approach using log-linear weighting will follow later.

Basically, this adds the `VectorContextSimilarity` class which plugs into the `DBTwoStepDisambiguator`, optionally replacing the `GenerativeContextSimilarity`. This class implements the ContextSimilarity interface using a store of vectors, which can be generated using models like Word2Vec or GloVe.

To use the VectorContextSimilarity, a `vectors.mem` file needs to be available in the spotlight data folder. This file can be generated by first training a Word2Vec model using Wiki2Vec, placing the resulting `enwiki-model-stemmed.w2c.syn0.csv`
`enwiki-model-stemmed.w2c.wordids.txt` files into `<spotlight-model-dir>/word2vec/`, and running `CreateSpotlightModel`.

### CODE REVIEW

File: conf/server.properties [hunk 1]
```diff
@@ -28,8 +28,8 @@ org.dbpedia.spotlight.language = English
 org.dbpedia.spotlight.language_i18n_code = en
 # Stop word list
 # An example can be downloaded from: http://spotlight.dbpedia.org/download/release-0.4/stopwords.en.list
-org.dbpedia.spotlight.data.stopWords.english = stopwords.en.list
-org.dbpedia.spotlight.data.stopWords.portuguese = /usr/local/spotlight/dbpedia_data/data/stopwords.pt.list
+org.dbpedia.spotlight.data.stopWords.english = data_quickstart/stopwords.en.list
```
Comment (tgalery) [inline conf/server.properties]: Maybe it would be good to add some comments in the file itself to explain why this folder is needed as some for of documentation

File: conf/server.properties [hunk 2]
```diff
@@ -37,10 +37,10 @@ org.dbpedia.spotlight.data.stopWords.portuguese = /usr/local/spotlight/dbpedia_d
 # Accepted values are LingPipeSpotter,WikiMarkupSpotter,AtLeastOneNounSelector,CoOccurrenceBasedSelector,NESpotter,OpenNLPNGramSpotter,OpenNLPChunkerSpotter,KeaSpotter
 # Some spotters may require extra files and config parameters. See org.dbpedia.spotlight.model.SpotterConfiguration
 org.dbpedia.spotlight.spot.spotters = LingPipeSpotter,WikiMarkupSpotter
-org.dbpedia.spotlight.spot.selectors =
+org.dbpedia.spotlight.spot.selectors = ShortSurfaceFormSelector
```
Comment (tgalery) [inline conf/server.properties]: Maybe add some comments explaining this choice here too ?

File: conf/server.properties (hunk 1 above)
Comment (phdowling) [inline conf/server.properties]: Sure, but afaik the whole server.properties file is only really applicable when the lucene backend is used, right?

Comment (tgalery) [inline conf/server.properties]: ha god knows!

File: conf/server.properties [hunk 3]
```diff
@@ -77,25 +77,25 @@ org.dbpedia.spotlight.spot.ahocorasick.surfaceforms=/data/spotlight/surfaceforms
...
 # Only used if one such disambiguator is loaded. Data is at: http://spotlight.dbpedia.org/download/release-0.5/candidate-index-full.tgz
-#org.dbpedia.spotlight.candidateMap.dir = dist/src/deb/control/data/usr/share/dbpedia-spotlight/index
+#org.dbpedia.spotlight.candidateMap.dir = dist/src/deb/control/data_quickstart/usr/share/dbpedia-spotlight/index
 
 
 #----- DISAMBIGUATION -------
 
 # List of disambiguators to load: Document,Occurrences,CuttingEdge,Default
-org.dbpedia.spotlight.disambiguate.disambiguators = CuttingEdge
+org.dbpedia.spotlight.disambiguate.disambiguators = Default,Document
```
Comment (tgalery) [inline conf/server.properties]: why did you change this ?

File: conf/server.properties [hunk 4]
```diff
@@ -110,14 +110,3 @@ jcs.default.cacheattributes.MaxObjects = 5000
 # Configuration for SparqlFilter
 org.dbpedia.spotlight.sparql.endpoint = http://dbpedia.org/sparql
 org.dbpedia.spotlight.sparql.graph = http://dbpedia.org
-
-# Topical classification properties
```
Comment (tgalery) [inline conf/server.properties]: I think that instead of removing these lines, it would be good to comment them. I remember some old GSoc project using that.  Maybe @dirkweissenborn has something to say on this ?

File: core/pom.xml [hunk 5]
```diff
@@ -230,10 +230,10 @@
             <artifactId>scalaz-core_2.10</artifactId>
         </dependency>
 
-        <dependency>
+        <!--dependency>
```
Comment (tgalery) [inline core/pom.xml]: you can just sack this off

File: core/pom.xml [hunk 6]
```diff
@@ -241,6 +241,23 @@
             <version>2.10.5</version>
         </dependency>
 
+        <dependency>
```
Comment (tgalery) [inline core/pom.xml]: The same dep is appearing twice (probs because of the rebases). You can keep one and clean up the whitespaces.
Also, you might consider using a more up to date version of breze, like the 0.11

File: core/pom.xml [hunk 7]
```diff
@@ -250,4 +267,4 @@
 
     </dependencies>
 
-</project>
+</project>
```
Comment (tgalery) [inline core/pom.xml]: weird EOF

File: core/src/main/scala/org/dbpedia/spotlight/db/SpotlightModel.scala [hunk 8]
```diff
@@ -53,17 +53,30 @@ object SpotlightModel {
...
+    } else if (new File(modelDataFolder, "context.mem").exists()) {
       MemoryStore.loadContextStore(new FileInputStream(new File(modelDataFolder, "context.mem")), tokenTypeStore, quantizedCountsStore)
-    else
+    } else{
       null
+    }
+
+    val vectorStore = if (new File(modelDataFolder, "vectors.mem").exists()){
+      MemoryStore.loadVectorStore(new FileInputStream(new File(modelDataFolder, "vectors.mem")))
+    }else{
```
Comment (tgalery) [inline core/src/main/scala/org/dbpedia/spotlight/db/SpotlightModel.scala]: else some spaces in the `else` statement to follow scala guidelines.

File: core/src/main/scala/org/dbpedia/spotlight/db/SpotlightModel.scala [hunk 9]
```diff
@@ -83,9 +96,14 @@ object SpotlightModel {
       case s: String => new SnowballStemmer(s)
     }
 
-    def contextSimilarity(): ContextSimilarity = contextStore match {
-      case store:MemoryContextStore => new GenerativeContextSimilarity(tokenTypeStore, contextStore)
-      case _ => new NoContextSimilarity(MathUtil.ln(1.0))
+    def contextSimilarity(): ContextSimilarity = {
+      if (vectorStore != null) {
+        new VectorContextSimilarity(vectorStore)
+      }else if (contextStore != null) {
```
Comment (tgalery) [inline core/src/main/scala/org/dbpedia/spotlight/db/SpotlightModel.scala]: spaces in the else

File: core/src/main/scala/org/dbpedia/spotlight/db/memory/MemoryStore.scala [hunk 10]
```diff
@@ -189,6 +198,11 @@ object MemoryStore {
     s
   }
 
+  def loadVectorStore(in: InputStream) = {
```
Comment (tgalery) [inline core/src/main/scala/org/dbpedia/spotlight/db/memory/MemoryStore.scala]: could you put a return type and just return the `load` statement into one line ?

File: core/src/main/scala/org/dbpedia/spotlight/db/memory/MemoryVectorStore.scala [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+  @transient
+  var resourceIdToVectorIndex: Map[Int, Int] = null
+
+  @transient
+  var tokenTypeIdToVectorIndex: Map[Int, Int] = null
+
+  override def size: Int = vectors.rows
+
+  def _lookup(id: Int) = {
```
Comment (tgalery) [inline core/src/main/scala/org/dbpedia/spotlight/db/memory/MemoryVectorStore.scala]: these function definitions seem a bit pythonic in style to me, you should use the scala's camel case and probably change the iiner _lookup so some other name `lookupItem` or something

File: core/src/main/scala/org/dbpedia/spotlight/db/memory/MemoryVectorStore.scala [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    // look up vector, if it isn't there, simply ignore the word
+    if(id != -1){
+      vectors(id, ::)
+    }else{
+      DenseVector.zeros[Float](vectors.cols).t
+    }
+  }
+
+  def _on_nil_index(string: String) = {
+    //println("Warning: token " + string + " not in dictionary! Lookup returning null vector.")
```
Comment (tgalery) [inline core/src/main/scala/org/dbpedia/spotlight/db/memory/MemoryVectorStore.scala]: Also it might be good to add logging statements where your print lines are, probably debug or info.

File: core/src/main/scala/org/dbpedia/spotlight/db/similarity/VectorContextSimilarity.scala [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+case class VectorContextSimilarity(memoryVectorStore: MemoryVectorStore) extends ContextSimilarity{
+
+  def vectorSimilarities(query: Seq[TokenType], candidates: Set[DBpediaResource]): mutable.Map[DBpediaResource, Double] = {
+    val contextScores = mutable.HashMap[DBpediaResource, Double]()
+
+    candidates.map( resource => {
+      contextScores.put(
+        resource,
+        // similarity of context and current resource
+        // TODO: use the whole context, or only surrounding n words?
```
Comment (tgalery) [inline core/src/main/scala/org/dbpedia/spotlight/db/similarity/VectorContextSimilarity.scala]: I was wondering if there is a nead way to capture of your TODO in terms of adding an extra argument (a function) that modifies the sequence of tokentypes. By default that function would be Identity, but we could pass different ones to measure the impact on performance

File: core/src/main/scala/org/dbpedia/spotlight/db/similarity/VectorContextSimilarity.scala [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        cosineSimilarity(
+          query.map(memoryVectorStore.lookup).reduceLeft(_ + _),
+          memoryVectorStore.lookup(resource)
+        )
+        //Random.nextDouble()
```
Comment (tgalery) [inline core/src/main/scala/org/dbpedia/spotlight/db/similarity/VectorContextSimilarity.scala]: remove comment

File: core/src/main/scala/org/dbpedia/spotlight/disambiguate/MultiThreadedDisambiguatorWrapper.scala [hunk 15]
```diff
@@ -34,12 +34,16 @@ package org.dbpedia.spotlight.disambiguate
  * limitations under the License.
  */
 
+
+
```
Comment (tgalery) [inline core/src/main/scala/org/dbpedia/spotlight/disambiguate/MultiThreadedDisambiguatorWrapper.scala]: remove spaces

File: core/src/main/scala/org/dbpedia/spotlight/util/MathUtil.scala [hunk 16]
```diff
@@ -1,5 +1,7 @@
 package org.dbpedia.spotlight.util
 
+import breeze.linalg.{DenseVector, Transpose}
+import breeze.numerics._
```
Comment (tgalery) [inline core/src/main/scala/org/dbpedia/spotlight/util/MathUtil.scala]: I think it would be better to import the specific classes that you use here.
Also add an extra line at the end of the file to remove the weird EOF.

File: eval/pom.xml [hunk 17]
```diff
@@ -137,4 +136,4 @@
 
     </dependencies>
 
-</project>
+</project>
```
Comment (tgalery) [inline eval/pom.xml]: weird eof

File: eval/src/main/scala/org/dbpedia/spotlight/evaluation/EvaluateParagraphDisambiguator.scala [hunk 18]
```diff
@@ -127,19 +134,22 @@ object EvaluateParagraphDisambiguator {
...
+                                 new DBBaselineDisambiguator(sfStore, resStore, candMapStore)
                                  //, new CuttingEdgeDisambiguator(factory),
                                  //new PageRankDisambiguator(factory)
                                 )
 
+
         val sources = List(//AidaCorpus.fromFile(new File("/home/pablo/eval/aida/gold/CoNLL-YAGO.tsv")),
-                           //MilneWittenCorpus.fromDirectory(new File("/home/pablo/eval/wikify/original"))
-                           PredoseCorpus.fromFile(new File("/home/pablo/eval/predose/predose_annotations.tsv")))
+                           MilneWittenCorpus.fromDirectory(new File("/home/dirk/gsoc/data/evaldata/data_gold/spotlight/wikify/original")))
```
Comment (tgalery) [inline eval/src/main/scala/org/dbpedia/spotlight/evaluation/EvaluateParagraphDisambiguator.scala]: This file path doesn't look sane, I wonder whether we could specify the location of the gold standard in the file input or something.

File: eval/src/main/scala/org/dbpedia/spotlight/evaluation/EvaluateParagraphDisambiguator.scala [hunk 19]
```diff
@@ -153,9 +163,8 @@ object EvaluateParagraphDisambiguator {
             }
         }
 
-        val occFilters = List(UriWhitelistFilter.fromFile(new File(conceptURIsFileName)),
-                              RedirectResolveFilter.fromFile(new File(redirectTCFileName)),
-                              noNils)
+        val occFilters = List()
```
Comment (tgalery) [inline eval/src/main/scala/org/dbpedia/spotlight/evaluation/EvaluateParagraphDisambiguator.scala]: I wonder what the role of these filters were previously.

File: index/src/main/scala/org/dbpedia/spotlight/db/CreateSpotlightModel.scala [hunk 20]
```diff
@@ -228,6 +229,13 @@ object CreateSpotlightModel {
     memoryIndexer.writeTokenOccurrences()
     memoryIndexer.writeQuantizedCounts()
 
+    val memoryVectorStoreIndexer = new MemoryVectorStoreIndexer(
+      new File(modelDataFolder, "/word2vec/enwiki-model-stemmed.w2c.syn0.csv"),
```
Comment (tgalery) [inline index/src/main/scala/org/dbpedia/spotlight/db/CreateSpotlightModel.scala]: wouldn't this look up the word2vec folder inside the root of the file system ?

File: pom.xml [hunk 21]
```diff
@@ -560,19 +560,29 @@
...
                 <artifactId>breeze-learn_2.10</artifactId>
                 <version>0.3</version>
-                <!--
                     License: Apache Software License, Version 2.0
-                -->
+            </dependency-->
+
+            <dependency>
+                <groupId>org.scalanlp</groupId>
+                <artifactId>breeze_2.10</artifactId>
```
Comment (tgalery) [inline pom.xml]: looks like the deps got duplicated here too

File: pom.xml [hunk 22]
```diff
@@ -560,19 +560,29 @@
...
+                <version>0.10</version>
             </dependency>
 
             <dependency>
                 <groupId>com.typesafe.akka</groupId>
                 <artifactId>akka-actor_2.10</artifactId>
-                <version>2.3.4</version>
+                <version>2.2.3</version>
```
Comment (tgalery) [inline pom.xml]: maybe the higher versions of the deps should be kept

**Comment by tgalery** [comment]
In addition to the comments I just made, it would be good to have test for:

(i) the functions you added in math-utils,
(ii) the classes you wrote

File: conf/server.properties (hunk 3 above)
Comment (phdowling) [inline conf/server.properties]: Honestly, I think this just happened during rebasing somehow. I didn't mean to change anything in that file, I think I'll just reverse this

File: core/src/main/scala/org/dbpedia/spotlight/db/similarity/VectorContextSimilarity.scala (hunk 13 above)
Comment (tgalery) [inline core/src/main/scala/org/dbpedia/spotlight/db/similarity/VectorContextSimilarity.scala]: Maybe just leave the todo for now

File: eval/src/main/scala/org/dbpedia/spotlight/evaluation/EvaluateParagraphDisambiguator.scala (hunk 19 above)
Comment (tgalery) [inline eval/src/main/scala/org/dbpedia/spotlight/evaluation/EvaluateParagraphDisambiguator.scala]: any thoughts on this @jodaiber ?

File: index/src/main/scala/org/dbpedia/spotlight/db/CreateSpotlightModel.scala [hunk 23]
```diff
@@ -245,9 +253,9 @@ object CreateSpotlightModel {
     } else {
       new LanguageIndependentTokenizer(Set[String](), stemmer, locale, tokenStore)
     }
-    val fsaDict = FSASpotter.buildDictionary(sfStore, tokenizer)
+    //val fsaDict = FSASpotter.buildDictionary(sfStore, tokenizer)
```
Comment (tgalery) [inline index/src/main/scala/org/dbpedia/spotlight/db/CreateSpotlightModel.scala]: Hey If you leave this commented out, we won't build the data structures for doing the spotting via the SFA. If you uncomment them do things break ?

Comment (phdowling) [inline index/src/main/scala/org/dbpedia/spotlight/db/CreateSpotlightModel.scala]: Good catch, I've put the line back in. It definitely should not break things, but actually this whole class is hard for me to test propery because there are so many dependencies. Does anyone have a setup in which they can easily build a new Spotlight model from scratch? I think it might be important to at least test this all the way through at some point.

**Comment by tgalery** [comment]
Looking good, but I got this error when trying to build the model

```
Loading vector dictionary!
[WARNING] 
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:293)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.FileNotFoundException: ./target_dir/model/word2vec/enwiki-model-stemmed.w2c.wordids.txt (No such file or directory)
	at java.io.FileInputStream.open(Native Method)
	at java.io.FileInputStream.<init>(FileInputStream.java:146)
	at scala.io.Source$.fromFile(Source.scala:90)
	at scala.io.Source$.fromFile(Source.scala:75)
	at scala.io.Source$.fromFile(Source.scala:80)
	at org.dbpedia.spotlight.db.MemoryVectorStoreIndexer.loadVectorDict(MemoryVectorStoreIndexer.scala:25)
	at org.dbpedia.spotlight.db.CreateSpotlightModel$.main(CreateSpotlightModel.scala:236)
	at org.dbpedia.spotlight.db.CreateSpotlightModel.main(CreateSpotlightModel.scala)
	... 6 more
[WARNING] thread Thread[ForkJoinPool-1-worker-7,5,org.dbpedia.spotlight.db.CreateSpotlightModel] was interrupted but is still alive after waiting at least 15000msecs
[WARNING] thread Thread[ForkJoinPool-1-worker-7,5,org.dbpedia.spotlight.db.CreateSpotlightModel] will linger despite being asked to die via interruption
[WARNING] NOTE: 1 thread(s) did not finish despite being asked to  via interruption. This is not a problem with exec:java, it is a problem with the running code. Although not serious, it should be remedied.
[WARNING] Couldn't destroy threadgroup org.codehaus.mojo.exec.ExecJavaMojo$IsolatedThreadGroup[name=org.dbpedia.spotlight.db.CreateSpotlightModel,maxpri=10]
java.lang.IllegalThreadStateException
	at java.lang.ThreadGroup.destroy(ThreadGroup.java:775)
	at org.codehaus.mojo.exec.ExecJavaMojo.execute(ExecJavaMojo.java:328)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)

```

gonna investigate later, as I need to go now

**Comment by tgalery** [comment]
looks like there are some harcoded bits

**Comment by phdowling** [comment]
Ah yes, the word2vec data of course needs to be in the right place to build the whole model. How should we pass the location in as a param, or where should we expect it to be?

**Comment by tgalery** [comment]
I think that model generation should be similar to model reading. At the moment, we supply a working directory with the raw counts, and the redirects instance types and disambiguations files. It would be nice to detect the presence of word2vec data there and generate the new vector model on that basis, if such file doesn't exist, then proceed in the old way. Note that the command arguments for the indexing specify a language and a locale, so you'd need to make the generation of the new model sensitive to that too.

**Comment by phdowling** [comment]
Okay, sounds reasonable. To be clear: the w2v model creating using gensim should not be included in spotlight, right (that's what we have Wiki2Vec for)? Other than that, yes the language sensitivity is a good point. Right now I have the path `new File(modelDataFolder, "word2vec/enwiki-model-stemmed.w2c.syn0.csv")` hardcoded in `CreateSpotlightModel`, this should definitely not be strictly enwiki, but I'm not sure we need to bind the filename to the language in any way at all, since we don't do this for the counts, redirects and disambiguations either (the supplied data is implicitly all the same language). Maybe we should just call the raw w2v files `word2vec-model-syn0.csv` and `word2vec-model-wordids.txt`? They get converted to storages and saved to `vectors.mem` by `CreateSpotlightModel`.

Can we address the issues with the model creating script itself as well? There seem to be non-working links within `index_db.sh`, I'm not sure how to properly run it to create a full model.

**Comment by tgalery** [comment]
Yeah that is fine, the modelDataFolder already contains language specific
info, e.g. stopwords, etc... so, you only need to make a convention on the
word2vec csv file (it could be just word2vec.csv) or something like that
and then just check for the existence of the file for indexing. This solves
the problem for this pr.

We would need to think of a tooling for generating those csv files though,
but that won't be your problem.

**Comment by phdowling** [comment]
I did document how I do this if I recall correctly. Might come in handy, and it's actually not too difficult. 

Could you comment on the missing links thing - is one of you guys looking to fix this, or should I see what I can do?

**Comment by tgalery** [comment]
Yeah, I remember you did, but that presupposes uses running wiki2vec on a
whole wikipedia, running your script to fix some issues, building a gensim
model, exporting it to csv and then placing that in the correct folder
before indexing. Am I right ?

If so, people will never use those models, ideally we would like to
automate that better, probably with the use of json-wikipedia.

As for missing links, you mean the issue that appeared in evaluation with
some of the entities not having corresponding vectors ? I think it's an
issue related to resolving redirects + the fact that we don't replace all
occurances of the words in a page by the link (e.g. Barack_Obama appears
linked only once). I think also that json wikipedia can really help with
that, if you want to poke this bear, I have a few ideas to discuss.

**Comment by phdowling** [comment]
Sorry, by missing links I actually meant the fact that the dbpedia data
download links in the index script are broken/outdated.

**Comment by tgalery** [comment]
I think there is a pr that fixess that, but that script teoretically should
be legacy once we have the spark system (although I'm not so sure when that
will happen), but for testing the indexing one can download those files
manually and run the indexer.
